ComponentKit/Core/CKComponentViewAttribute.h

```diff
@@ -74,52 +114,205 @@ struct CKComponentViewAttribute {
-  CKBoxedValue(uint32_t v) noexcept : __actual(@(v)) {};
-  CKBoxedValue(int64_t v) noexcept : __actual(@(v)) {};
-  CKBoxedValue(uint64_t v) noexcept : __actual(@(v)) {};
-  CKBoxedValue(long v) noexcept : __actual(@(v)) {};
-  CKBoxedValue(unsigned long v) noexcept : __actual(@(v)) {};
-  CKBoxedValue(float v) noexcept : __actual(@(v)) {};
-  CKBoxedValue(double v) noexcept : __actual(@(v)) {};
-  CKBoxedValue(SEL v) noexcept : __actual([NSValue valueWithPointer:v]) {};
-  CKBoxedValue(std::nullptr_t v) noexcept : __actual(nil) {};
-
-  // Any objects go here
-  CKBoxedValue(id obj) noexcept : __actual(obj) {};
-
-  // Define conversions for common Apple types
-  CKBoxedValue(CGRect v) noexcept : __actual([NSValue valueWithCGRect:v]) {};
-  CKBoxedValue(CGPoint v) noexcept : __actual([NSValue valueWithCGPoint:v]) {};
-  CKBoxedValue(CGSize v) noexcept : __actual([NSValue valueWithCGSize:v]) {};
-  CKBoxedValue(UIEdgeInsets v) noexcept : __actual([NSValue valueWithUIEdgeInsets:v]) {};
-
-  operator id () const {
-    return __actual;
-  };
+namespace CK {
+  namespace ViewAttribute {
 
-private:
-  id __actual;
+    // Singleton object to signify non-ID-type object. For internal use only.
+    extern id nonIDObject;
 
-};
+    /**
+     An abstract base type for the value type. This is a hack to store multiple templated objects in a non
+     templated collection.
+     */
+    struct ValueBase
+    {
+      ValueBase(std::string typeName) : _typeName(typeName) {}
+      ValueBase(const ValueBase &) = delete;
+      ~ValueBase() {}
+
+      virtual operator id() const = 0;
+      virtual size_t hash() const noexcept = 0;
+      virtual BOOL isEqualTo(const ValueBase &) const = 0;
+      virtual void performSetter(id object, SEL setter) const = 0;
+
+      const std::string typeName() const
+      {
+        return _typeName;
+      }
+
+    protected:
+      const std::string _typeName;
+    };
+
+    /**
+     Templated type used to store an underlying value.
+     */
+    template<typename T> struct Value : ValueBase
+    {
+      Value(const T &value) : ValueBase(typeid(Value<T>).name()), _value(value) {}
+      Value(const Value<T> &) = delete;
+      ~Value() {}
+
+      operator const T() const
+      {
+        return _value;
+      }
+
+      operator id() const
+      {
+        return nonIDObject;
+      }
+
+      template<typename U> operator const Value<U>() const
+      {
+        return Value<U>(_value);
+      }
+
+      size_t hash() const noexcept
+      {
+        return CK::hash(_value);
+      }
+
+      BOOL isEqualTo(const ValueBase &other) const
+      {
+        return this->typeName() == other.typeName() && CK::is_equal<T>()(_value, static_cast<const Value<T> &>(other)._value);
+      }
+
+      void performSetter(id object, SEL setter) const
+      {
+        const auto setterIMP = (void (*)(id, SEL, T))[object methodForSelector:setter];
+        setterIMP(object, setter, _value);
```

Comment:
Thinking about this more, by using IMPs, don't we disallow message forwarding entirely? That's probably OK, since if you need forwarding you can use the block syntax. This should be faster than invocations.

I'd like to add some type-checking here for DEBUG mode though. I'd grab the type signature from the object and verify that the type encoding for _value matches `@encode(T)`? Also add an assertion if setterIMP is NULL and no-op so we don't crash without a good debug message?